### PR TITLE
m2m

### PR DESCRIPTION
**Pull request by xdavidz**
this is the first version of data-driver m2m in polling mode. Still working on intr and better code structure. Please feel free to add any comments.

```
root@xsjyliu50:~# xbutil m2mtest
INFO: Found total 1 card(s), 1 are usable
HBM[0] -> HBM[1] M2M bandwidth: 9898.69 MB/s
HBM[0] -> HBM[2] M2M bandwidth: 9905.97 MB/s
HBM[0] -> HBM[3] M2M bandwidth: 9906.74 MB/s

```

> Changed ioctl name to copy_bo as a generic interface;
check in execbuf when SSV3 m2m subdev is on, bug_on the error case;
tested p2p case and wait for Soren's code, this should not block this PR to be merged.
enhanced xrt_cu API, enabled enable_intr, disable_intr and clear_intr for hls CU.

### CODE REVIEW

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+#define	M2M_INFO(m2m, fmt, arg...)	\
+	xocl_info(&(m2m)->m2m_pdev->dev, fmt "\n", ##arg)
+#define	M2M_DBG(m2m, fmt, arg...)	\
+	xocl_dbg(&(m2m)->m2m_pdev->dev, fmt "\n", ##arg)
+
+/* This is the real register map for the copy bo CU */
+struct start_copybo_cu_cmd {
+  uint32_t src_addr_lo;      /* low 32 bit of src addr */
+  uint32_t src_addr_hi;      /* high 32 bit of src addr */
+  uint32_t src_bo_hdl;       /* src bo handle, cleared by driver */
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c:40]: What does "cleared by driver" mean?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+  uint32_t dst_addr_lo;      /* low 32 bit of dst addr */
+  uint32_t dst_addr_hi;      /* high 32 bit of dst addr */
+  uint32_t dst_bo_hdl;       /* dst bo handle, cleared by driver */
+  uint32_t size;             /* size in bytes */
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c:44]: This register should be size in bus width. In current design, the bus width is 512 bits (64 bytes). Please define a macro "#define KDMA_DATA_WIDTH 64".

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+	struct xrt_cu *xcu = &m2m->m2m_cu;
+	struct start_copybo_cu_cmd cmd;
+
+	cmd.src_addr_lo = (uint32_t)src_paddr;
+	cmd.src_addr_hi = (src_paddr >> 32) & 0xFFFFFFFF;
+	cmd.dst_addr_lo = (uint32_t)dst_paddr;
+	cmd.dst_addr_hi = (dst_paddr >> 32) & 0xFFFFFFFF;
+	cmd.src_bo_hdl = src_bo_hdl;
+	cmd.dst_bo_hdl = dst_bo_hdl;
+	cmd.size = size;
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c:68]: cmd.size = size / KDMA_DATA_WIDTH;

Comment (mamin506) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c:68]: Oops, the argument size is already calculated with bus width in xocl_bo.c. Please add a comment to let people know cmd.size is not in bytes.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c (hunk 2 above)
Comment (mamin506) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c:44]: Already have "KDMA_BLOCK_SIZE" in ert.h, no need to define a new macro.
Please only correct this comment.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c (hunk 3 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c:68]: Yes, fixed it with additional comments

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c (hunk 2 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c:44]: adjusted the comments

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c (hunk 1 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/m2m.c:40]: removed it, copied from ert.h

**Comment by xdavidz** [comment]
> In today's code, copy BO CU cmd will be sent down to driver, where it has enough information to tell if this copy is b/w two BOs on the same board (m2m) or on different boards (p2p). It seems that your implementation will just blindly send it down to m2m subdev if it's available for copy. Not sure how to we handle P2P copy case? Is P2P copy tested?

Great cache, forgot to handle the p2p and added back into the code. Haven't tested yet. Will figure out how to verify that.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_bo.c [hunk 4]
```diff
@@ -1329,3 +1329,70 @@ int xocl_usage_stat_ioctl(struct drm_device *dev, void *data,
...
+		XOCL_DRM_GEM_OBJECT_PUT_UNLOCKED(obj);
+		return -EINVAL;
+	}
+
+	*paddrp = xobj->mm_node->start + off;
+	XOCL_DRM_GEM_OBJECT_PUT_UNLOCKED(obj);
+	return 0;
+}
+
+int xocl_m2m_bo_ioctl(struct drm_device *dev, void *data,
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_bo.c:1364]: if you added back p2p bo copy, this function name should be changed to xocl_copy_bo_ioctl()? it's not just "m2m" any more, right?

**Comment by maxzhen** [comment]
> 
> 
> > In today's code, copy BO CU cmd will be sent down to driver, where it has enough information to tell if this copy is b/w two BOs on the same board (m2m) or on different boards (p2p). It seems that your implementation will just blindly send it down to m2m subdev if it's available for copy. Not sure how to we handle P2P copy case? Is P2P copy tested?
> 
> Great cache, forgot to handle the p2p and added back into the code. Haven't tested yet. Will figure out how to verify that.

This test requires two devices and DSV should have existing test cases. You might want to ask Sid/Uday for it.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_bo.c (hunk 4 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_bo.c:1364]: AI for me, once we discuss the P2P case, I would like to adjust the name to a proper name that we agree.

**Comment by xdavidz** [comment]
> > > In today's code, copy BO CU cmd will be sent down to driver, where it has enough information to tell if this copy is b/w two BOs on the same board (m2m) or on different boards (p2p). It seems that your implementation will just blindly send it down to m2m subdev if it's available for copy. Not sure how to we handle P2P copy case? Is P2P copy tested?
> > 
> > 
> > Great cache, forgot to handle the p2p and added back into the code. Haven't tested yet. Will figure out how to verify that.
> 
> This test requires two devices and DSV should have existing test cases. You might want to ask Sid/Uday for it.

Asked Chien-wei and Sid for many version of P2P cases. Combined their case together and verify P2P works.
However, the code path will always choose execBuf instead of xoclCopyBO. Will need to discuss this topic with the team. I have set up a meeting on Monday.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/mb_scheduler.c [hunk 5]
```diff
@@ -4238,6 +4238,9 @@ static int convert_execbuf(struct xocl_dev *xdev, struct drm_file *filp,
 	if (scmd->opcode != ERT_START_COPYBO)
 		return 0;
 
+	/* If SSV3 M2M subdev is enabled, execbuf API should not be called for copybo */
+	BUG_ON(M2M_CB(xdev));
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/mb_scheduler.c:4242]: sorry, this bug on has to be changed to a failure return with error msg. jsut realized that it's coming from user space, which should not be allowed to crash driver. my bad.

File: src/runtime_src/core/common/drv/cu_hls.c [hunk 6]
```diff
@@ -122,13 +122,89 @@ static void cu_hls_check(void *core, struct xcu_status *status)
 	status->num_ready = ready_reg;
 }
 
+static void cu_hls_enable_intr(void *core, u32 intr_type)
+{
+	struct xrt_cu_hls *cu_hls = core;
+	u32 *base_addr = cu_hls->vaddr;
```
Comment (mamin506) [inline src/runtime_src/core/common/drv/cu_hls.c:128]: Please use the same pointer type in cu_hls_enable_intr() and cu_hls_disable_intr().

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/mb_scheduler.c (hunk 5 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/mb_scheduler.c]: fixed, thanks.

**Pull request merged**